src/script/standard.cpp

```diff
@@ -89,21 +89,45 @@ static constexpr bool IsSmallInteger(opcodetype opcode)
     return opcode >= OP_1 && opcode <= OP_16;
 }
 
+/** Valid number of keys in a multisig are [1,20]. */
+static constexpr bool IsValidMultisigKeyCount(unsigned int n_keys)
+{
+    return n_keys > 0 && n_keys <= MAX_PUBKEYS_PER_MULTISIG;
+}
+
+static bool GetMultisigKeyCount(const CScript& script, CScript::const_iterator& it, unsigned int& count, opcodetype opcode, valtype data)
+{
+    if (IsSmallInteger(opcode)) {
+        count = CScript::DecodeOP_N(opcode);
+    } else {
+        try {
+            count = CScriptNum(data, true).getint();
+        } catch (const scriptnum_error&) {
+            return false;
+        }
+    }
+    if (!IsValidMultisigKeyCount(count)) return false;
+
+    return true;
+}
+
 static bool MatchMultisig(const CScript& script, unsigned int& required, std::vector<valtype>& pubkeys)
 {
     opcodetype opcode;
     valtype data;
+    unsigned int keys;
+
     CScript::const_iterator it = script.begin();
     if (script.size() < 1 || script.back() != OP_CHECKMULTISIG) return false;
 
-    if (!script.GetOp(it, opcode, data) || !IsSmallInteger(opcode)) return false;
-    required = CScript::DecodeOP_N(opcode);
+    if (!script.GetOp(it, opcode, data) || !GetMultisigKeyCount(script, it, required, opcode, data)) return false;
     while (script.GetOp(it, opcode, data) && CPubKey::ValidSize(data)) {
         pubkeys.emplace_back(std::move(data));
     }
-    if (!IsSmallInteger(opcode)) return false;
-    unsigned int keys = CScript::DecodeOP_N(opcode);
+    if (!GetMultisigKeyCount(script, it, keys, opcode, data)) return false;
```

Comment:
Is this necessary? Seems you already `return false` on this condition at L123?

Reply:
On L123 i get the count of the required number of keys (the `M` of the `M of N`) and here the total number of keys (the `N` of the `M of N`)